internal/snmp/wrapper.go

```diff
@@ -165,18 +165,8 @@ func (gs *GosnmpWrapper) SetAgent(agent string) error {
 	}
 
 	switch u.Scheme {
-	case "tcp":
-		gs.Transport = "tcp"
-	case "tcp4":
-		gs.Transport = "tcp4"
-	case "tcp6":
-		gs.Transport = "tcp6"
-	case "udp":
-		gs.Transport = "udp"
-	case "udp4":
-		gs.Transport = "udp4"
-	case "udp6":
-		gs.Transport = "udp6"
+	case "tcp", "tcp4", "tcp6", "udp", "udp4", "udp6":
+		gs.Transport = u.Scheme
 	default:
 		return fmt.Errorf("unsupported scheme: %v", u.Scheme)
 	}
```

Comment:
This got me thinking, why even bother checking it? This block could be replaced with `gs.Transport = u.Scheme` and gosnmp will handle it further when connecting.

Reply:
Hi and thank you for you reviews! This is a pleasant experience! 

I thought the same thing but I am not sure of the side-effects that the remaining schemes like unix-sockets have. I am not sure if this plays well with windows or if it would play well with non-privileged users on Unix.
After all, the only schemes we do not support with the above code are ` "unix", "unixgram" and "unixpacket" `. 
Usually I prefer to just expose everything the library has to offer and let the user handle the aftermath, but in this case I am a little intimidated by the size of the project here. 
I think it would be better to limit the connection possibilities to the "well-known" ones. 

What do you think of that?

Reply:
My opinion is still to let gosnmp handle it. I'm calling @reimda in to hear his opinion.

Reply:
I did a little digging and I think I know what the intention behind the check in gosnmpwrapper was: 
gosnmp will happily take arbitrary values as connetion scheme and not complain about them. The first time the code can fail due to a wrong connection scheme is when `gs.Connect` in `plugins/input/snmp.go` is called.

However, the error raised by gosnmp is quite descriptive of what is going on:
```agent upd4://test: setting up connection: error establishing connection to host: dial upd4:test: unknown network upd```
Additionally, this is error is handeled here (line 576 ongoing):
```
if err := gs.Connect(); err != nil {
   return nil, fmt.Errorf("setting up connection: %w", err)
}
```
So indeed: There should be no problem in letting gosnmp handle the error and just feeding it with anything in`u.Scheme`.
Additionally this is more future-proof... maybe gosnmp will add some more schemes later on.

Reply:
Correct indeed, that's what I meant.

Reply:
Currently I am looking into why the schemes `ip`, `ip4` and `ip6` always seem to fail and return UnknownNetworkError. 
I traced it down to `dial.go` in the go-sources line 175. 
```
167 func parseNetwork(ctx context.Context, network string, needsProto bool) (afnet string, proto int, err error) {
168     i := last(network, ':')
169     if i < 0 { // no colon
170         switch network {
171         case "tcp", "tcp4", "tcp6":
172         case "udp", "udp4", "udp6":
173         case "ip", "ip4", "ip6":
174             if needsProto {
175                 return "", 0, UnknownNetworkError(network)
176             }
177         case "unix", "unixgram", "unixpacket":
178         default:
179             return "", 0, UnknownNetworkError(network)
180         }
181         return network, 0, nil
182     }
```
Of course: As soon as you specify a port, you also need to specify the protocol! However, the error is not really clear on whats happening. 

Since we are using port 161 by default for snmp, it is clear that the aformentiones schemes will not work. And in fact, they don't really make any sense to me... I really don't know why go-snmp offers them in the first place. 
For the sockets as connection-schemes: 
A socket does not need a port. Unfortunately, specifying something like `unix://test:123` will also result in an `UnknownNetworkError`. I think this can be quite confusing for the user. 

Maybe, after all, it does make sense to limit the connection-schemes to udp{4,6} and tcp{4,6}. Or we have to implement checks in `snmp.go` to ensure correct formatting of the schemes. 
Maybe this is the reason why the connection-schemes were limited in the first place. 
Otherwise, mis-configuration may result in confusing error-messages for the user. 

Penny for your thoughts :)

Reply:
Okay, you convinced me :laughing: Fine for me to keep it as is.

Reply:
I added some documentation, so the next guy is spared of this discovery-process :)
Thanks for your input. Greatly appreciated!